### ISSUE TITLE
possible interaction problem with seccomp and disabled selinux in systemd 231

### ISSUE THREAD

**Issue by webczat** [systemd/systemd#3962]
Hello, I would like to report a bug, possibly a regression, as it happens only with systemd-231 from what I can tell.
My distribution is archlinux, running on my server, with a manually compiled linux kernel with heavily modified configuration.
My systemd does have compiled in support for selinux, and I have all necessary selinux support tools, but selinux itself is disabled in /etc/selinux/config.
The problem is that the system starts, with exception of few services. Those are key services: systemd-journald, systemd-timesyncd, systemd-resolved and systemd-networkd. Logs, that is dmesg, show that control processes of all those services were killed using signal SIGSYS.
I believe this is a bug resulting from some interaction that happens only with selinux compiled in but disabled in kernel, and seccomp filters. This is what I tested to confirm the above:
- I replaced temporarily my kernel with original distro one, there was no change, services didn't start.
- on my laptop, where there is the same distro but without selinux support compiled in the kernel and without selinux in other packages, everything works well.
- even if there is no policy, changing selinux mode to permissive in the server configuration mitigates the bug and services start. I could of course restart them too, but when I then typed "echo 1 > /sys/fs/selinux/disable" and tried to restart systemd-journald, it was killed by SIGSYS as before.
- if selinux is disabled in kernel, but service files of affected services are modified not to use seccomp, the bug does not appear too.

That is currently all I know. I know this bug did not appear in systemd-230, even with selinux support compiled in.

**Comment by evverx** [comment]
> Logs, that is dmesg, show that control processes of all those services were killed using signal SIGSYS

Can you provide the output of the

``` sh
dmesg | grep type=1326
```

?

(You should see something like `audit: type=1326 audit(1471109203.980:49): auid=4294967295 uid=0 gid=0 ses=4294967295 pid=280 comm="(sh)" exe="/usr/lib/systemd/systemd" sig=31 arch=c000003e syscall=41 compat=0 ip=0x7fbe90427757 code=0x0`)

**Comment by webczat** [comment]
well, because at the time of the error journald was disabled and nothing
enabled audit logging, maybe auditd, I do not have such things here. but
I may be able to trigger it in a controlled way now
I tried. I have auditd running so I was reading audit logs:
and they show so
type=SECCOMP msg=audit(1471263423.920:286): auid=4294967295 uid=0 gid=0
ses=4294
967295 pid=8456 comm="systemd-timesyn"
exe="/usr/lib/systemd/systemd-timesyncd"
sig=31 arch=c000003e syscall=165 compat=0 ip=0x7f239101df5a code=0x0
not sure if the same syscall happens with all problematic services...

**Comment by mbiebl** [comment]
Might be a duplicate of #3882

**Comment by webczat** [comment]
no, not confirmed. seccomp is enabled here and works properly. The audit
log fragment shows, I believe, that seccomp just blocked a syscall not
found in filters. But, based on my findings as to when this bug does not
appear, it seems this syscall may be somehow related to support of
selinux in the kernel. this bug will never trigger with enabled selinux
in kernel, nor with disabled selinux in systemd ./configure.

**Comment by evverx** [comment]
Well,

```
sig=31 arch=c000003e syscall=165
```

means:
- `seccomp` is enabled
- a `mount`-call was blocked by filter

I think this was fixed by https://github.com/SELinuxProject/selinux/commit/32773a99b1f0cf2b

**Comment by webczat** [comment]
so you are actually suggesting libselinux bug that has been fixed
recently? great. why that didn't happen in systemd 230? did it have less
restrictive seccomp or no seccomp at all, maybe?

**Comment by evverx** [comment]
`SyscallFilter` was introduced by 40093ce5d

```
$ git describe --contains 40093ce5dd444
v231~201
```

Try to remove `@mount`. Does it work?

**Comment by webczat** [comment]
Okay. of course it does work when @mount is removed.

**Comment by evverx** [comment]
Are you using https://aur.archlinux.org/packages/libselinux/ ?
I think @fishilico can help to backport https://github.com/SELinuxProject/selinux/commit/32773a99b1f0cf2b

**Comment by webczat** [comment]
Yes, I am using that package. of course not sure if it is better to
backport the change, or to just... wait for new libselinux release, like
it does not matter for me as workarounds do exist, as we saw before.

**Comment by fishilico** [comment]
Hello, as I was able to reproduce this issue on my test system (by disabling SELinux after having it installed), I backported the fixes in the libselinux package I am maintaining in the AUR (https://aur.archlinux.org/cgit/aur.git/commit/?h=libselinux&id=5f8a4aa9ef53c19e192342eb4108c6c405d1e101).
Thanks for your clear messages about what was going on, they are very much appreciated. Please let me know if there still exist issues after this.

**Comment by webczat** [comment]
No problem, I was just doing much testing and had some ideas.
So, thanks for backporting, it seems that everything works correctly
now. So, waiting for official next version of libselinux.

**Comment by evverx** [comment]
@fishilico , thanks!

For the record:
http://pkgs.fedoraproject.org/cgit/rpms/libselinux.git/commit/?id=fd9d18361dccc4bbf8c185f0dc1b4af7e9521924

https://anonscm.debian.org/git/selinux/libselinux.git/commit/?id=f8713974c7e925bc82506a090eade13193f57783

**Comment by poettering** [comment]
It's really strange that libselinux tries to mount things on its own in this case. It's good to know this got fixed upstream already.

**Comment by webczat** [comment]
I ever saw any problems, but does that mean it confused processes? or at
some point we had a proc fs mounted few times in mount table?

**Comment by poettering** [comment]
well, libselinux decided to do system setup stuff even though it's just a library that people call to do transitions and stuff. It shouldn't as side-effect init your system, possibly in a different way than the actual init system.
